### DISCUSSION TITLE
Consuming external Kafka topics/messages

### DISCUSSION THREAD

**Discussion by konigbach** [roadrunner-server/roadrunner#1802] [Q&A] [answered]
Hello! I am a little bit lost regarding the consumer kafka topics part: I understand that you defined the topics to be consumed in the config file but, which php class is going to handle it? Let's say I have a topic 'users' with different messages ('user-created', 'user-deleted'...). How do I set 'this topic is going to be handled by this php-class'? I know how to work with Spiral jobs (based on names) but consuming external messages is a mystery to me (currently, I'm dispatching the messages from a python service). Thank you! :)

**Comment by rustatian** [comment]
Hey @konigbach 👋 
@wolfy-j @msmakouz Could you please help?
From the RR part, you'll always have a `topic` in the received message.

**Reply by konigbach** [reply]
I have this in app/config/queue.php (I also tried it in rr.yaml)
![image](https://github.com/roadrunner-server/roadrunner/assets/17541795/bf3f2c1b-63a1-4f59-8e48-9d5cc49b5347)
but I'm missing (or not understanding) where to say: hey, every time you consume a 'users' topic, go this handler-class.

**Comment by msmakouz** [comment, marked as answer]
@konigbach Hi. Do you pass the job name in your service that pushes the job to the queue? The name does not necessarily have to be the class name, you can use any string.

And you can register a standalone handler for each topic.  For example:

```php
// app/config/queue.php
return [
    'registry' => [
        'handlers' => [
            'users' => App\Endpoint\Job\UsersJob::class,
             // ...
        ],
    ],
];
```

And push jobs specifying this name (instead of the class name). This is an example in PHP, but your Python service should have something similar.

```php
$queue->push('users', ['name' => 'value']);
```

**Reply by konigbach** [reply]
Thanks a lot! :)

**Reply by konigbach** [reply]
Hey @msmakouz! Sorry for disturbing you. I know have
```   'pipelines' => [
        'default' => [
            'connector' => new KafkaCreateInfo(
                name: 'kafka',
                consumerOptions: new ConsumerOptions(
                    topics: ['userstest'],
                ),
            ),
            'consume' => true,
        ],
    ],

    'registry' => [
        'handlers' => [
            'userstest' => App\Endpoint\Job\UserCreatedJob::class,
        ],
    ],
```
The message is consumed but it says:
`"msg":"jobs protocol error","error":"ReflectionException: Class \"deduced_by_rr\" does not exist
`    
Can you please give me a hint? Thanks!

**Reply by msmakouz** [reply]
@konigbach Hi, in your error, **deduced_by_rr** is the default name set by RoadRunner if it is not provided. This means you are not passing the name. In this case, your task can be solved in a different way. You can add an interceptor where you can get the topic name, and then pass the topic name further as the task name. Interceptor example:

```php
namespace App\Interceptor;

use Spiral\Core\CoreInterceptorInterface;
use Spiral\Core\CoreInterface;

final class TopicInterceptor implements CoreInterceptorInterface
{
    private const ROADRUNNER_DEFAULT_NAME = 'deduced_by_rr';

    public function process(string $name, string $action, array $parameters, CoreInterface $core): mixed
    {
        if ($name !== self::ROADRUNNER_DEFAULT_NAME) {
            return $core->callAction($name, $action, $parameters);
        }

        $topic = $parameters['queue'];

        // We skip handler existence check here, add it if you need it
        return $core->callAction($topic, $action, $parameters);
    }
}
```

In this interceptor, we check if the task name is the default one set by RoadRunner (if not, we bypass this interceptor and proceed with further execution using $core->callAction(...)). Next, we retrieve the topic name from $parameters['queue'] and pass it as the first parameter to the $core->callAction call, instead of the task name. This way, your handler will be found based on it.

The interceptor needs to be registered in the configuration file queue.php:

```php
'interceptors' => [
    'consume' => [
        \App\Interceptor\TopicInterceptor::class,
    ],
],
```

More about interceptors: https://spiral.dev/docs/queue-interceptors/current/en#interceptor-for-consume

I've made a small example: https://github.com/msmakouz/kafka-topic-interceptor-example
Interceptor: https://github.com/msmakouz/kafka-topic-interceptor-example/blob/master/app/app/src/Interceptor/TopicInterceptor.php
Config: https://github.com/msmakouz/kafka-topic-interceptor-example/blob/master/app/app/config/queue.php
RR config: https://github.com/msmakouz/kafka-topic-interceptor-example/blob/master/app/.rr.yaml
And the console command I used to test the job push: https://github.com/msmakouz/kafka-topic-interceptor-example/blob/master/app/app/src/Endpoint/Console/PushCommand.php (I specifically set the name to 'deduced_by_rr', which is the same as the default set by RoadRunner).

**Reply by konigbach** [reply]
@msmakouz I was exactly looking for that. Thanks a lot for the help and the quick and complete responses!

**Reply by konigbach** [reply]
@msmakouz Hello again. I'm having some issues while working on this: with your project, apparently everything works fine. I tried many many times trying to reproduce it.
```
version: '3'
rpc:
    listen: 'tcp://127.0.0.1:6001'

jobs:
    pool:
        num_workers: 2
        max_worker_memory: 100
    consume: ["kafka"]
    pipelines:
      kafka:
        driver: kafka
        config:
          auto_create_topics_enable: true
          consumer_options:
            topics: [ "new" ]

kafka:
  brokers: [ "127.0.0.1:9092" ]

server:
    command: 'php app.php'
    relay: pipes     
```
The interceptor
```
final class Topic extends JobHandler
{
    public function invoke(
        mixed $payload,
        LoggerInterface $logger,
    ): void {
        $logger->info('entered');
    }
}

```
and the config file
```
    'default' => 'rr-kafka',

    'connections' => [
        'sync' => [
            'driver' => 'sync',
        ],
        'rr-memory' => [
            'driver' => 'roadrunner',
            'pipeline' => 'memory',
        ],
        'rr-kafka' => [
            'driver' => 'roadrunner',
            'pipeline' => 'kafka',
        ],
    ],

    'pipelines' => [
        'memory' => [
            'connector' => new MemoryCreateInfo(
                name: 'memory',
            ),
            'consume' => true,
        ],
        'kafka' => [
            'connector' => new KafkaCreateInfo(
                name: 'kafka',
            ),
            'consume' => true,
        ],
    ],

    'registry' => [
        'handlers' => [
            'new' => App\Endpoint\Job\Topic::class,
        ],
    ],

    'interceptors' => [
        'consume' => [
            TopicInterceptor::class,
        ],
    ],
```

It consumes the messages but the interceptor is not being used (therefore, the DeducedByRr missing class appears, which make sense). Am I missing something?

**Reply by msmakouz** [reply]
@konigbach Hello. What version of Spiral Framework are you using? If it's below 3.10.1, try updating to 3.10.1, as the RetryPolicyInterceptor was fixed in this version:
https://github.com/spiral/framework/pull/1024
Also, could you please show the code of the interceptor? The configuration seems correct. It would be great if you could upload an example of your non-working code to the repository so that it can be viewed and reproduced, thanks.

**Reply by konigbach** [reply]
- The version is 3.10.1

![image](https://github.com/roadrunner-server/roadrunner/assets/17541795/eac7253c-c2eb-4096-99f5-b02c076a6ffc)
- This is the code inside the interceptor: dispatch another job with the name of the kafka topic to handle it. This is working in the project you made and share with me. I tried in brand new projects and for some reason the interceptor is not 'intercepting' (I debug a bit the container and apparently is there). I will make a small project to share with you with the interceptor example. Thanks!
```
<?php

declare(strict_types=1);

namespace App\Application\Bootloader;

use Psr\Log\LoggerInterface;
use Spiral\Core\CoreInterceptorInterface;
use Spiral\Core\CoreInterface;
use Spiral\Queue\QueueConnectionProviderInterface;

final class TopicInterceptor implements CoreInterceptorInterface
{
    public function __construct(
        private readonly QueueConnectionProviderInterface $queueConnectionProvider,
        private readonly LoggerInterface $logger,
    ) {
    }

    public function process(
        string $controller,
        string $action,
        array $parameters,
        CoreInterface $core,
    ): mixed {
        $this->logger->info('TopicInterceptor');

        if ($controller !== 'deduced_by_rr') {
            return $core->callAction($controller, $action, $parameters);
        }

        $this->logger->info('Not a spiral job');

        $topic = $parameters['queue'];
        $this->queueConnectionProvider->getConnection('sync')->push($topic, $parameters);

        // We could add a handler existence check here
        return null;
    }
}
```

**Reply by konigbach** [reply]
From scratch: https://github.com/konigbach/spiral-interceptor.git

**Reply by msmakouz** [reply]
@konigbach You need to substitute the job handler's name (from **deduced_by_rr** to **phones** - from topic name) in the interceptor and invoke further execution with the new name by calling **return $core->callAction($topic, $action, $parameters);.** As it was in my example. In your example, further execution proceeds with the name **deduced_by_rr** and accordingly, it does not find such a handler in the config file.

**Reply by konigbach** [reply]
@msmakouz
Let me rephrase the problem: the flow doesn't go to the interceptor
```
class TopicInterceptor implements CoreInterceptorInterface
{
    public function process(
        string $controller,
        string $action,
        array $parameters,
        CoreInterface $core
    ): mixed {
        dd(123);
    }
}
```

If I put this in your project, it throws the dd. If I put it in a brand new project, it doesn't. My issue is that the interceptor is not there for some reason. Thanks again for your patience :)

**Reply by konigbach** [reply]
@msmakouz Hello! I have found that the version of roadrunner package is different. 
I think this is the commit that broke it: https://github.com/spiral/roadrunner-bridge/pull/81/files
With spiral/roadrunner-bridge 3.0.2 it works. With spiral/roadrunner-bridge >= 3.1.0 it doesn't

**Reply by msmakouz** [reply]
@konigbach Looks like a bug, we will check it. Specify the 3.0.2 version in the composer for now.

**Reply by TheLevti** [reply]
Is there a result on whether its a bug? We unfortunately still struggle with this.

**Reply by rustatian** [reply]
@msmakouz

**Reply by msmakouz** [reply]
@TheLevti Should be fixed in this release:
https://github.com/spiral/roadrunner-bridge/releases/tag/3.3.1
If the problem is still reproducible, please create a simple example in the GitHub repository.

**Reply by TheLevti** [reply]
Thank you very much, will do.

**Comment by suityou01** [comment]
I am having problems just getting the basic config set up. Should I create a new discussion topic?

```config
version: '3'

rpc:
  listen: tcp://127.0.0.1:6001

server:
  command: "php consumer.php"

#http:
#  address: "0.0.0.0:8080"
kafka:
  addrs: 127.0.0.1:9092

jobs:
  num_pollers: 1
  pipeline_size: 100
  pool:
    num_workers: 1
    max_jobs: 0
    allocate_timeout: 60s
    destroy_timeout: 60s
  consume: ["kafka"]
  pipelines:
    kafka:
      driver: kafka
      config:
        priority: 1
        topic: feeds.feeds.unit
        brokers: [ broker:9092 ]
        consumer_options:
          topics: [ "feeds", "feeds.feeds.unit", "^[a-zA-Z0-9._-]+$" ]
          consume_regexp: true
logs:
  level: error
```

Error "Did not find expected key by parsing a block mapping"

NB The example has a key called topic at the jobs.pipelines.queue-name.config level which is confusing
e.g.

```yaml
# Kafka jobs driver
#
# This option is required to use Kafka driver. Addrs can contain any number of addresses separated by comma (127.0.0.1:9092,127.0.0.1:9093,...)
kafka:
  addrs: 127.0.0.1:9092

jobs:
  num_pollers: 10
  pipeline_size: 100000
  pool:
    num_workers: 10
    max_jobs: 0
    allocate_timeout: 60s
    destroy_timeout: 60s

  pipelines:
    test-local-6:
      # Driver name
      #
      # This option is required
      driver: kafka

      # Driver's configuration
      #
      # Should not be empty
      config:

        # Pipeline priority
        #
        # If the job has priority set to 0, it will inherit the pipeline's priority. Default: 10.
        priority: 1

        # Topic name: https://kafka.apache.org/intro#intro_concepts_and_terms
        #
        # This option is required and should not be empty.
        topic: test-1
```
So we have to specify a topic here (single topic, which one??)
And for the consumer we have to specify a topics key

How does this work?

**Reply by rustatian** [reply]
@suityou01 Yeah, would be nice to have a separate discussion topic 😄
